server/interaction/npc.go

```diff
@@ -0,0 +1,123 @@
+package interaction
+
+import (
+	"context"
+	"errors"
+
+	"github.com/joshprzybyszewski/cribbage/logic/scorer"
+	"github.com/joshprzybyszewski/cribbage/model"
+	"github.com/joshprzybyszewski/cribbage/utils/rand"
+)
+
+const (
+	Dumb   model.PlayerID = `DumbNPC`
+	Simple model.PlayerID = `SimpleNPC`
+	Calc   model.PlayerID = `CalculatedNPC`
+)
+
+var (
+	ErrUnknownNPCType = errors.New(`unknown NPC type`)
+)
+
+var npcs = map[model.PlayerID]npc{
+	Dumb:   &dumbNPC{},
+	Simple: &simpleNPC{},
+	Calc:   &calculatedNPC{},
+}
+
+var _ Player = (*NPCPlayer)(nil)
+
+type NPCPlayer struct {
+	HandleActionCallback func(ctx context.Context, a model.PlayerAction) error
+	player               npc
+	id                   model.PlayerID
+}
+
+// NewNPCPlayer creates a new NPC with specified type
+func NewNPCPlayer(pID model.PlayerID, cb func(ctx context.Context, a model.PlayerAction) error) (Player, error) {
+	p, ok := npcs[pID]
+	if !ok {
+		return nil, ErrUnknownNPCType
+	}
+	return &NPCPlayer{
+		player:               p,
+		id:                   pID,
+		HandleActionCallback: cb,
+	}, nil
+}
+
+func (npc *NPCPlayer) ID() model.PlayerID {
+	return npc.id
+}
+
+func (npc *NPCPlayer) NotifyBlocking(b model.Blocker, g model.Game, s string) error {
+	a, err := npc.buildAction(b, g)
+	if err != nil {
+		return err
+	}
+	// TODO find a better way to solve this problem...
```

Comment:
the real answer is that if a game is "checked out" of the database and somebody is working on it, then we shouldn't allow anyone else to touch that game until the other process finishes it work with it. I'm ok leaving a `time.Sleep` in here until we can get the transactional work done :#